source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -103,12 +103,13 @@ void Filter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callb
   decoding_state_.setDecoderFilterCallbacks(callbacks);
   const Envoy::StreamInfo::FilterStateSharedPtr& filter_state =
       callbacks.streamInfo().filterState();
-  if (!filter_state->hasData<ExtProcLoggingInfo>(ExtProcLoggingInfoName)) {
-    filter_state->setData(ExtProcLoggingInfoName, std::make_shared<ExtProcLoggingInfo>(),
+  if (!filter_state->hasData<ExtProcLoggingInfo>(callbacks.filterConfigName())) {
+    filter_state->setData(callbacks.filterConfigName(),
+                          std::make_shared<ExtProcLoggingInfo>(config_->filterMetadata()),
```

Comment:
From the proto and change log, I think this filter metadata should be injected into the dynamic metadata rather than the filter state?

Reply:
I agree it's very confusing and I'm trying to track down the different uses and threads that discussed this.
I was mainly following the guidelines from [this comment](https://github.com/envoyproxy/envoy/issues/4929#issuecomment-519783954):
> * data that has to be read/write from multiple filters, and need a sort of query configuration, (such as data consumed by RBAC) uses dyanmic metadata
> * data that has clear purpose to a known filters, such as envoy.tcp_proxy.cluster, envoy.network.upstream_server_name, uses filter state.

but I guess this may also be out-of-date.

Two things:
1. If information was passed via filter-state and now it is being passed via some other mechanism, this should at least be runtime-guarded. Filter-state is (maybe unfortunately) not officially part of the API, and this change may break some other filter that uses these values (but so does renaming the key name).
2. I'd like to highlight is that the added metadata is not really _dynamic_ in it's nature, but I guess that passing it over dynamic structures is done because they are available.

Reply:
There is no official guidance. But typically I will use them in this way: the filter state will used to store the data generated by filter and defined by c++ code. And dynamic metadata will be used to prapagate the API/xds/proto/config related things.
I personally prefer filter state because it usually faster.
---

Specific to this comment, I have no strong point on should using dynamic metadata or filter state.

But the document and the change log tell the end users the new `filter_metadata` will be added to stream info's dynamic metadata, then our implementation should obey this point.

Reply:
So the only user of this hardcoded filter state is Google. Sadly and fortunately, this filter state was not even documented in our docs. So changing it without runtime guard is okay but after this change, we are making sure that the name of this state is explicitly documented in our docs. So any future changes to the name of this state will require runtime guard.